### PR TITLE
Improve the logging to allow an audit trail for restconf

### PR DESCRIPTION
**Pull request by gcampbell512**
For security reasons a device should be capable of generating an audit log of all modifications made to a device.

This change allows the logging down to the leaf level of any modifications made to the configuration via restconf.

### CODE REVIEW

**Review by tony-vanderpeet — APPROVED** [review]

File: fcgi.c [hunk 1]
```diff
@@ -174,7 +174,7 @@ handle_http (void *arg)
 {
     FCGX_Request *request = (FCGX_Request *) arg;
     char *rpath, *path, *length, *if_match, *if_none_match, *if_modified_since, *if_unmodified_since;
-    char *server_name, *server_port;
+    char *server_name, *server_port, *remote_addr, *remote_user;
```
Comment (carlgsmith) [inline fcgi.c:177]: remote_port (REMOTE_USER is not a CGI environment variable)

File: internal.h [hunk 2]
```diff
@@ -86,13 +98,18 @@ void fcgi_stop (void);
...
                const char *if_match, const char *if_none_match,
                const char *if_modified_since, const char *if_unmodified_since,
-               const char *server_name, const char *server_port, const char *data, int length);
+               const char *server_name, const char *server_port,
+               const char *remote_addr, const char *remote_user,
+               const char *data, int length);
 void rest_shutdown (void);
 void yang_library_create (sch_instance *schema);
 void restconf_monitoring_create (sch_instance *schema);
+bool restconf_logging_test_flag (int flag);
```
Comment (carlgsmith) [inline internal.h:111]: these do not need a "restconf" prefix ... just logging_init, logging_shutdown, logging_test_flag

File: internal.h [hunk 3]
```diff
@@ -86,13 +98,18 @@ void fcgi_stop (void);
 /* Rest */
 extern bool rest_use_arrays;
 extern bool rest_use_types;
-gboolean rest_init (const char *path);
+gboolean rest_init (const char *path, const char *logging);
```
Comment (carlgsmith) [inline internal.h:101]: I dont think this needs to be done in rest.c ... the logging init/shutdown can be done in main.c

File: rest.c [hunk 4]
```diff
@@ -142,6 +151,118 @@ restconf_error (int status, rest_e_tag error_tag)
...
+        while (rnode && rnode->children)
+            rnode = rnode->children;
+
+        params.root_path = apteryx_node_path (rnode);
+        params.remote_user = remote_user;
+        params.remote_addr = remote_addr;
+        params.key = key;
+        params.rc = rc;
+
+        if (tree)
```
Comment (carlgsmith) [inline rest.c:229]: why not just log the tree output ... this could generate thousands of log entries

File: internal.h [hunk 5]
```diff
@@ -48,6 +48,17 @@ extern bool verbose;
...
+{
+    LOG_NONE                    = 0,
+    LOG_POST                    = (1 << 0),  /* Log POST requests */
+    LOG_PUT                     = (1 << 1),  /* Log PUT requests */
+    LOG_PATCH                   = (1 << 2),  /* Log PATCH requests */
+    LOG_DELETE                  = (1 << 3),  /* Log DELETE requests */
+    LOG_GET                     = (1 << 4),  /* Log GET requests */
+    LOG_HEAD                    = (1 << 5),  /* Log HEAD requests */
+} logging_flags;
+
```
Comment (carlgsmith) [inline internal.h:61]: a global "int logging" would be simpler ... so then the code could just check "if (logging & LOG_POST)" etc

File: rest.c [hunk 6]
```diff
@@ -242,6 +364,9 @@ rest_api_search (const char *path, const char *if_none_match, const char *if_mod
     rc = apteryx_json_search (path, &data);
 
   exit:
+    if ((flags & FLAGS_RESTCONF))
```
Comment (carlgsmith) [inline rest.c:367]: should not be restconf specific (that flag will be removed eventually) ... but should be "if (logging)"

File: rest.c [hunk 7]
```diff
@@ -142,6 +151,118 @@ restconf_error (int status, rest_e_tag error_tag)
...
+        key = "DELETE";
+
+        params.root_path = NULL;
+        params.remote_user = remote_user;
+        params.remote_addr = remote_addr;
+        params.key = key;
+        params.rc = rc;
+
+        if (tree)
+            g_node_traverse (tree, G_IN_ORDER, G_TRAVERSE_LEAVES, -1, log_deleted_leafs,
```
Comment (carlgsmith) [inline rest.c:258]: ditto .. just log the tree output

File: rest.c [hunk 8]
```diff
@@ -1179,12 +1322,17 @@ rest_api (req_handle handle, int flags, const char *rpath, const char *path,
...
-rest_init (const char *path)
+rest_init (const char *path, const char *logging)
 {
     struct sysinfo info;
     struct timespec monotime;
     struct timespec monotime_raw;
 
+    if (logging)
+    {
+        restconf_logging_init (path, logging);
```
Comment (carlgsmith) [inline rest.c:1333]: call from main.c not here

File: fcgi.c (hunk 1 above)
Comment (gcampbell512) [reply fcgi.c:179]: REMOTE_USER is part of the CGI 1.1 standard, and supported in some fcgi implementations. If REMOTE_USER is not defined in the env block the remote_user will be set to "unknown"

File: internal.h (hunk 2 above)
Comment (gcampbell512) [reply internal.h:111]: Done

File: internal.h (hunk 3 above)
Comment (gcampbell512) [reply internal.h:101]: Done

File: rest.c (hunk 4 above)
Comment (gcampbell512) [reply rest.c:229]: A proper audit log requires each individual change to the system configuration to be logged.

File: internal.h (hunk 5 above)
Comment (gcampbell512) [reply internal.h:61]: Done

File: rest.c (hunk 6 above)
Comment (gcampbell512) [reply rest.c:367]: Done

File: rest.c (hunk 8 above)
Comment (gcampbell512) [reply rest.c:1333]: Done

File: rest.c (hunk 7 above)
Comment (gcampbell512) [reply rest.c:258]: A proper audit log requires each individual change to the system configuration to be logged.

File: internal.h [hunk 9]
```diff
@@ -90,9 +102,17 @@ gboolean rest_init (const char *path);
...
-               const char *server_name, const char *server_port, const char *data, int length);
+               const char *server_name, const char *server_port,
+               const char *remote_addr, const char *remote_user,
+               const char *data, int length);
 void rest_shutdown (void);
 void yang_library_create (sch_instance *schema);
 void restconf_monitoring_create (sch_instance *schema);
 
+/* Logging */
+extern logging_flags logging;
```
Comment (carlgsmith) [inline internal.h:113]: "int logging;" they are flags .. not a single enum

File: rest.c (hunk 4 above)
Comment (carlgsmith) [reply rest.c:229]: what would you miss with the logging of the tree output. In fact the method above is more misleading as it suggests that there were individual set operations. Even from the CLI a single atomic change in configuration would often involve a tree set.

**Review by carlgsmith — APPROVED** [review]

**Pull request merged**